backend/app/checkpoint.py

```diff
@@ -1,147 +1,112 @@
-                    thread_id,
-                    datetime.fromisoformat(thread_ts),
-                ):
-                    return CheckpointTuple(
-                        config,
-                        loads(value[0]),
-                        {
-                            "configurable": {
-                                "thread_id": thread_id,
-                                "thread_ts": value[1],
-                            }
-                        }
-                        if value[1]
-                        else None,
-                    )
-            else:
-                if value := await conn.fetchrow(
-                    "SELECT checkpoint, thread_ts, parent_ts FROM checkpoints WHERE thread_id = $1 ORDER BY thread_ts DESC LIMIT 1",
-                    thread_id,
-                ):
-                    return CheckpointTuple(
-                        {
-                            "configurable": {
-                                "thread_id": thread_id,
-                                "thread_ts": value[1],
-                            }
-                        },
-                        loads(value[0]),
-                        {
-                            "configurable": {
-                                "thread_id": thread_id,
-                                "thread_ts": value[2],
-                            }
-                        }
-                        if value[2]
-                        else None,
-                    )
-
-    async def aput(self, config: RunnableConfig, checkpoint: Checkpoint) -> None:
-        thread_id = config["configurable"]["thread_id"]
-        async with get_pg_pool().acquire() as conn:
-            await conn.execute(
-                """
-                INSERT INTO checkpoints (thread_id, thread_ts, parent_ts, checkpoint)
-                VALUES ($1, $2, $3, $4) 
-                ON CONFLICT (thread_id, thread_ts) 
-                DO UPDATE SET checkpoint = EXCLUDED.checkpoint;""",
-                thread_id,
-                datetime.fromisoformat(checkpoint["ts"]),
-                datetime.fromisoformat(checkpoint.get("parent_ts"))
-                if checkpoint.get("parent_ts")
-                else None,
-                pickle.dumps(checkpoint),
+    async def ensure_setup(self) -> None:
+        """Ensure the instance is set up before use."""
+        if not self._setup_complete:
+            await self.setup()
+            self._setup_complete = True
+
+    async def setup(self) -> None:
+        """Internal setup method."""
+        try:
+            conninfo = (
+                f"postgresql://{os.environ['POSTGRES_USER']}:"
+                f"{os.environ['POSTGRES_PASSWORD']}@"
+                f"{os.environ['POSTGRES_HOST']}:"
+                f"{os.environ['POSTGRES_PORT']}/"
+                f"{os.environ['POSTGRES_DB']}"
+            )
+
+            conn = AsyncConnectionPool(
```

Comment:
RuntimeWarning: opening the async pool AsyncConnectionPool in the constructor is deprecated and will not be supported anymore in a future release. Please use `await pool.open
()`, or use the pool as context manager using: `async with AsyncConnectionPool(...) as pool: